### ISSUE TITLE
Errno::EINVAL while running capybara tests in rails

### ISSUE THREAD

**Issue by will89** [puma/puma#1582]
### Steps to reproduce

1) Use `Timecop` to freeze time pretty far back in time (e.g. 1970)

2) Use capybara visit function to go to a url

### Expected behavior

Capybara should be able to go to the url.

### Actual behavior

We get the following stack trace.
```
Error in reactor loop escaped: Invalid argument (Errno::EINVAL)
puma-3.11.0/lib/puma/reactor.rb:29:in `select'
puma-3.11.0/lib/puma/reactor.rb:29:in `run_internal'
puma-3.11.0/lib/puma/reactor.rb:154:in `block in run_in_thread'
```

### System configuration

2.5.1
5.1.6
3.11.0
macOS High Sierra Version 10.13.2

### Investigation
It seems pretty easy in hindsight, but the stack trace was being generated because the `@sleep_for` value was too large, https://github.com/puma/puma/blob/master/lib/puma/reactor.rb#L126. In our test runs, this value was `1526858764.467461`.

I'm not sure if there is necessarily anything to fix here, but this was confusing for quite some time. Some possible ideas.
1) Raise a more informative error. The `man select` on my system does give two reasons for `EINVAL` so this is ambiguous to developers without really digging in.
2) Would it make sense to set a max value in this calculation, https://github.com/puma/puma/blob/master/lib/puma/reactor.rb#L282?
3) Is there a more useful way to share this?

**Comment by mileslane** [comment]
This looks very similar.  I don't know how to get more helpful background info on how to reproduce the problem.  Please let me know how I can help.

  * capybara (3.9.0)
  * capybara-selenium (0.0.6)
  * chromedriver-helper (2.1.0)
  * rspec (3.8.0)
  * rspec-core (3.8.0)
  * rspec-expectations (3.8.2)
  * rspec-instafail (1.0.0)
  * rspec-mocks (3.8.0)
  * rspec-rails (3.8.0)
  * rspec-support (3.8.0)
  * puma (3.12.0)

Error in reactor loop escaped: Invalid argument (Errno::EINVAL)
ruby-2.3.7@clear_value_plus/gems/puma-3.12.0/lib/puma/reactor.rb:126:in 'select'
ruby-2.3.7@clear_value_plus/gems/puma-3.12.0/lib/puma/reactor.rb:126:in 'run_internal'
ruby-2.3.7@clear_value_plus/gems/puma-3.12.0/lib/puma/reactor.rb:251:in 'block in run_in_thread'

**Comment by shepmaster** [comment]
@mileslane are you using Timecop or ActiveSupport's TimeHelpers in the test reporting the warning?

**Comment by mileslane** [comment]
Examples:

spec/models/user.rb
68:      Timecop.freeze('2015-04-01 12:00:00')
78:      Timecop.freeze('2015-04-01 12:00:00')
213:      Timecop.freeze do
225:      Timecop.freeze do

spec/models/reporter_spec.rb
17:    Timecop.travel(2018, 3, 25, 8, 00) { check_alert(false, 23.hours.ago) }
23:    Timecop.travel(2018, 3, 23, 8, 1) { check_alert(true, 2.hours.ago) }

**Comment by dzirtusss** [comment]
### Alternative quick-fix tip - don't travel to future, travel to past only
After some investigations, in our case these errors are happening only when dates are stubbed **to future date**. When stubbed **to past date** there is no errors (guess because delta is negative 😄 ), so..... just refactored everything to past.

**Comment by shepmaster** [comment]
> don't travel to future, travel to past only

@dzirtusss the [original comment says](https://github.com/puma/puma/issues/1582#issue-324993207):

> Use Timecop to freeze time pretty far back in time (e.g. 1970)

The OP is already going to the past.

**Comment by joethrive** [comment]
Upgraded from capybara (2.18.0) to capybara (3.10.1) and started receiving this error.  Rolling back to 2.18.0 solved the problem for me.  My current version of timecop is 0.9.1 and puma is 3.12.0.

**Comment by mileslane** [comment]
Downgrading capybara to 2.18.0 worked for me as well.  Thanks!
timecop (0.9.1)
puma (3.12.0)

**Comment by Hirurg103** [comment]
The same issue after upgrading to `capybara (3.11.1)`

**Comment by elizabrock** [comment]
I experienced this issue as well while upgrading testing gems.  I believe that this issue appears when upgrading capybara because version 3.0.0+ [changed how Puma is used](https://github.com/teamcapybara/capybara/blob/master/History.md#version-300), and additionally something between the 2.x version we were on and the current version of Capybara changed the default server from Webrick to Puma.

I found that the issue with time traveling wasn't traveling to the future vs. the past but in fact seemed to be based on traveling to different years.  Our suite time traveled to a variety of recent past dates (2014, 2015, etc) and all of those tests had issues post-upgrade.  Merely changing them all to time travel to the current year (2018) fixed the issues.  I did not dig in further, but hopefully these tidbits help!

P.S. This issue occurs with `Time.travel`, not just `Time.freeze`.

**Comment by mileslane** [comment]
@2rba has a patch that fixes this problem for me.  Can we please get this added to a release ASAP?  I have just tested with the patch applied.

rspec 3.8
rails 5.2.2.1
capybara 3.15.0
puma 3.12.1
timecop 0.9.1

**Comment by JonKernPA** [comment]
Another option:

```
# features/support/env.rb
Capybara.server = :webrick
```
* from [here](https://github.com/rspec/rspec-rails/issues/1908).

**Comment by nateberkopec** [comment]
Need a PR with tests.

**Comment by cjlarose** [comment]
The patch from @2rba worked for me in a Puma 3.x application, but throws errors in a Puma 4.x application.

The original `Errno::EINVAL` issue appears to be caused by passing very large values to the `select(2)` system call. But Puma 4.x stopped using the same call directly in (via #1728).

You can verify this manually by patching `Puma::Reactor` to set `@sleep_for` to a big number

```ruby
require 'puma/reactor'

module Puma
  class Reactor
    def calculate_sleep
      @sleep_for = 100_000_000_000.0
    end
  end
end
```

This causes Puma 3.x to emit the `EINVAL` errors, but Puma 4.x works just fine.

I think it is safe to say that this issue is effectively resolved by #1728. Upgrade to Puma 4.x to fix this issue in your applications.
